mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/MapperResultProducer.java

```diff
@@ -0,0 +1,107 @@
+ * from DAO methods.
+ *
+ * <p>For example, this could be used to substitute a 3rd-party future implementation for {@link
+ * CompletionStage}:
+ *
+ * <pre>
+ * public class CustomFutureProducer implements MapperResultProducer {
+ *   ...
+ * }
+ * </pre>
+ *
+ * <p>Producers are registered at construction time with {@link
+ * MapperBuilder#withResultProducers(MapperResultProducer...)}:
+ *
+ * <pre>
+ * InventoryMapper inventoryMapper = new InventoryMapperBuilder(session)
+ *     .withResultProducers(new CustomFutureProducer())
+ *     .build();
+ * </pre>
+ *
+ * DAO methods can then use the new type:
+ *
+ * <pre>
+ * &#64;Dao
+ * public interface ProductDao {
+ *   &#64;Select
+ *   CustomFuture&lt;Product&gt; findById(UUID productId);
+ * }
+ * </pre>
+ *
+ * See the javadocs of the methods in this interface for more explanations.
+ */
+public interface MapperResultProducer {
+
+  /**
+   * Checks if this producer can handle a particular result type.
+   *
+   * <p>This will be invoked at runtime to select a producer: if a DAO method declares a return type
+   * that is not supported natively, then the mapper generates an implementation which, for every
+   * invocation, iterates through all the producers <em>in the order that they were registered</em>,
+   * and picks the first one where {@code canProduce()} returns true.
+   *
+   * @param resultType the DAO method's declared return type. If checking the top-level type is
+   *     sufficient, then {@link GenericType#getRawType()} should do the trick. If you need to
+   *     recurse into the type arguments, call {@link GenericType#getType()} and use the {@code
+   *     java.lang.reflect} APIs.
+   */
+  boolean canProduce(GenericType<?> resultType);
+
+  /**
+   * Executes the statement generated by the mapper, and converts the result to the expected type.
+   *
+   * <p>This will be executed at runtime, every time the DAO method is called.
+   *
+   * @param statement the statement, ready to execute: the mapper has already bound all the values,
+   *     and set all the necessary attributes (consistency, page size, etc).
+   * @param context the context in which the DAO method is executed. In particular, this is how you
+   *     get access to the {@linkplain MapperContext#getSession() session}.
+   * @param entityHelper if the type to produce contains a mapped entity (e.g. {@code
+   *     ListenableFuture<Product>}), an instance of the helper class to manipulate that entity. In
+   *     particular, {@link EntityHelper#get(GettableByName) entityHelper.get()} allows you to
+   *     convert rows into entity instances. If the type to produce does not contain an entity, this
+   *     will be {@code null}.
+   * @return the object to return from the DAO method. This must match the type that this producer
+   *     was selected for, there will be an unchecked cast at runtime.
+   */
+  @SuppressWarnings("TypeParameterUnusedInFormals")
+  <EntityT> Object execute(
```

Comment:
Maybe add nullability annotations in this interface?

Reply:
Right, I forgot.